flink-runtime/src/main/java/org/apache/flink/runtime/net/RedirectingSslHandler.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelInboundHandlerAdapter;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.ByteToMessageDecoder;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.DefaultFullHttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.FullHttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpRequestDecoder;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseEncoder;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpVersion;
+import org.apache.flink.shaded.netty4.io.netty.handler.ssl.SslHandler;
+import org.apache.flink.shaded.netty4.io.netty.util.ReferenceCountUtil;
+
+import org.jboss.netty.handler.codec.http.HttpHeaders;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/** SSL handler which automatically redirects Non-SSL requests to SSL address. */
+public class RedirectingSslHandler extends ByteToMessageDecoder {
+	protected final Logger log = LoggerFactory.getLogger(getClass());
+
+	/** the length of the ssl record header (in bytes). */
+	private static final int SSL_RECORD_HEADER_LENGTH = 5;
+
+	private final String confRedirectBaseUrl;
+	private final CompletableFuture<String> redirectBaseUrl;
+	private final SSLEngineFactory sslEngineFactory;
+
+	public RedirectingSslHandler(
+		String confRedirectHost,
+		CompletableFuture<String> redirectBaseUrl,
+		SSLEngineFactory sslEngineFactory) {
+		this.confRedirectBaseUrl = "https://" + confRedirectHost + ":";
+		this.redirectBaseUrl = redirectBaseUrl;
+		this.sslEngineFactory = sslEngineFactory;
+	}
+
+	@Override
+	protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out) {
+		if (in.readableBytes() < SSL_RECORD_HEADER_LENGTH) {
+			return;
+		}
+		if (SslHandler.isEncrypted(in)) {
+			handleSsl(context);
+		} else {
+			context.pipeline().replace(this, "http-decoder", new HttpRequestDecoder());
+			context.pipeline().addAfter("http-decoder", "redirecting-non-ssl", newNonSslHandler());
+		}
+	}
+
+	private void handleSsl(ChannelHandlerContext context) {
+		SslHandler sslHandler = null;
+		try {
+			sslHandler = new SslHandler(sslEngineFactory.createSSLEngine());
+			context.pipeline().replace(this, "ssl", sslHandler);
+			sslHandler = null;
+		} finally {
+			// Since the SslHandler was not inserted into the pipeline the ownership of the SSLEngine was not
+			// transferred to the SslHandler.
+			if (sslHandler != null) {
+				ReferenceCountUtil.safeRelease(sslHandler.engine());
+			}
+		}
+	}
+
+	private ChannelHandler newNonSslHandler() {
+		return new ChannelInboundHandlerAdapter() {
```

Comment:
I would make this a proper class, right now we're creating an entirely new class every time a non-ssl request comes in.